### PR TITLE
gh-83845: Add tests for operator module

### PR DESCRIPTION
**Pull request by hugovk**
Builds on @tirkarthi's GH-18537, and tests `operator.index` with `int`, `float`, `Fraction`, `Decimal`, `None`.

Bumps coverage from 97% to 100%.

* Issue: gh-83845

### CODE REVIEW

File: Lib/test/test_operator.py [hunk 1]
```diff
@@ -508,6 +510,44 @@ def __getitem__(self, other): return 5  # so that C is a sequence
...
         self.assertEqual(operator.iconcat  (c, c), "iadd")
 
+    def test_iconcat_without_getitem(self):
+        operator = self.module
+
+        msg = "'int' object can't be concatenated"
+        with self.assertRaisesRegex(TypeError, msg):
+            operator.iconcat(1, 0.5)
+
+    def test_index(self):
```
Comment (JelleZijlstra) [inline Lib/test/test_operator.py:520]: Looking at the code, it seems like this one has some additional behavior in the C version that isn't reflected in the Python version.

1. Returning a subclass of int raises a DeprecationWarning:
```pycon
>>> class X:
...     def __index__(self): return True
... 
>>> operator.index(X())
<stdin>-3:1: DeprecationWarning: __index__ returned non-int (type bool).  The ability to return an instance of a strict subclass of int is deprecated, and may be removed in a future version of Python.
  operator.index(X())
1
```

But the Python version returns True.

2. If a subclass of int overrides `__index__`, the C version ignores it:
```
>>> class X(int):
...     def __index__(self): return 42
... 
>>> index(X())
42
>>> operator.index(X())
0
```

Arguably these are bugs, but I am not sure it's worth complicating the Python implementation for.

File: Lib/test/test_operator.py [hunk 2]
```diff
@@ -533,6 +573,13 @@ def __length_hint__(self):
         with self.assertRaises(LookupError):
             operator.length_hint(X(LookupError))
 
+        class Y: pass
```
Comment (JelleZijlstra) [inline Lib/test/test_operator.py:576]: Here the Python and C versions will behave differently if `__length_hint__` returns an instance of a subclass of int.

Python:
```
>>> class X:
...     def __length_hint__(self): return True
... 
>>> operator.length_hint(X())
True
```

C:
```
>>> class X:
...     def __length_hint__(self): return True
... 
>>> operator.length_hint(X())
1
```

Arguably a bug but again I don't know if it's worth complicating the code for.

**Review by JelleZijlstra — APPROVED** [review]
These tests look good. A close reading of the C code found a number of edge cases where the Python and C versions differ. Arguably those are bugs but I am not sure they are worth complicating the code for.

**Comment by hugovk** [comment]
Thanks for the review and finding those discrepancies. I agree it sounds reasonable to leave the implementations as is, we can always update it in the future if needed.

File: Lib/test/test_operator.py [hunk 3]
```diff
@@ -508,6 +510,44 @@ def __getitem__(self, other): return 5  # so that C is a sequence
...
+            operator.index(None)
+
+    def test_not_(self):
+        operator = self.module
+        class C:
+            def __bool__(self):
+                raise SyntaxError
+        self.assertRaises(TypeError, operator.not_)
+        self.assertRaises(SyntaxError, operator.not_, C())
+        self.assertFalse(operator.not_(5))
```
Comment (serhiy-storchaka) [inline Lib/test/test_operator.py:546]: BTW, I prefer to write this like
```python
self.assertIs(operator.not_(5), False)
```
to ensure that the result is boolean.

File: Lib/test/test_operator.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/test/test_operator.py:520]: It is a known bug, and there is an open issue for this.

**Pull request merged**